tests/rest/media/v1/test_media_storage.py

```diff
@@ -201,36 +202,49 @@ class _TestImage:
     ],
 )
 class MediaRepoTests(unittest.HomeserverTestCase):
-
+    test_image: ClassVar[_TestImage]
     hijack_auth = True
     user_id = "@test:user"
 
     def make_homeserver(self, reactor: MemoryReactor, clock: Clock) -> HomeServer:
 
-        self.fetches = []
+        self.fetches: List[
+            Tuple[
+                "Deferred[Tuple[bytes, Tuple[int, Dict[bytes, List[bytes]]]]]",
+                str,
+                str,
+                Optional[QueryParams],
+            ]
+        ] = []
 
         def get_file(
             destination: str,
             path: str,
             output_stream: BinaryIO,
-            args: Optional[Dict[str, Union[str, List[str]]]] = None,
+            args: Optional[QueryParams] = None,
+            retry_on_dns_fail: bool = True,
             max_size: Optional[int] = None,
-        ) -> Deferred:
-            """
-            Returns tuple[int,dict,str,int] of file length, response headers,
-            absolute URI, and response code.
-            """
+            ignore_backoff: bool = False,
+        ) -> Deferred[Tuple[int, Dict[bytes, List[bytes]]]]:
+            """A mock for MatrixFederationHttpClient.get_file."""
 
-            def write_to(r):
+            def write_to(
+                r: Tuple[bytes, Tuple[int, Dict[bytes, List[bytes]]]]
+            ) -> Tuple[int, Dict[bytes, List[bytes]]]:
                 data, response = r
                 output_stream.write(data)
                 return response
 
-            d = Deferred()
-            d.addCallback(write_to)
+            d: Deferred[Tuple[bytes, Tuple[int, Dict[bytes, List[bytes]]]]] = Deferred()
+            # Callbacks can _change_ the value held within a resolved Deferred(!!).
+            # The return type of addCallback reflects this...
+            d_after_callback = d.addCallback(write_to)
+            # ... even though d.addCallback returns d!
+            assert d_after_callback is d  # type: ignore[comparison-overlap]
             self.fetches.append((d, destination, path, args))
-            return make_deferred_yieldable(d)
+            return make_deferred_yieldable(d_after_callback)
```

Comment:
This was a huge surprise to me and feels like a footgun waiting to happen. Or am I overreacting?

Reply:
Callbacks chain, this is a fundamental part of Twisted and is well documented.

https://docs.twistedmatrix.com/en/stable/core/howto/defer.html#visual-explanation is the best resource I know of which explains this.